### ISSUE TITLE
-Wtautological-compare in arch/powerpc/boot/main.c

### ISSUE THREAD

**Issue by shenki** [ClangBuiltLinux/linux#212]
```
  BOOTCC  arch/powerpc/boot/main.o
arch/powerpc/boot/main.c:111:18: warning: array comparison always evaluates to a constant
      [-Wtautological-compare]
        if (_initrd_end > _initrd_start) {
                        ^
```

**Comment by tpimh** [comment]
I'm not familiar with PowerPC. What is BOOTCC? Couldn't find anything relevant.

**Comment by shenki** [comment]
```
$ git grep BOOTCC
arch/powerpc/boot/Makefile:#    To this end we need to define BOOTCC, etc, as the tools
arch/powerpc/boot/Makefile:    BOOTCC := $(CROSS32_COMPILE)gcc
arch/powerpc/boot/Makefile:    BOOTCC := $(CC)
arch/powerpc/boot/Makefile:BOOTCFLAGS   += -isystem $(shell $(BOOTCC) -print-file-name=include)
arch/powerpc/boot/Makefile:quiet_cmd_bootcc = BOOTCC  $@
arch/powerpc/boot/Makefile:      cmd_bootcc = $(BOOTCC) -Wp,-MD,$(depfile) $(BOOTCFLAGS) -c -o $@ $<
arch/powerpc/boot/Makefile:      cmd_bootas = $(BOOTCC) -Wp,-MD,$(depfile) $(BOOTAFLAGS) -c -o $@ $<
```

It's the compiler used to compile the boot wrapper for a powerpc kernel. The wrapper is built 32-bit for the 64-bit kernel in some situations, hence the different compiler line.

```
$ export ARCH=powerpc CROSS_COMPILE=powerpc64le-linux-gnu- CC=clang-7 
$ make powernv_defconfig
$ make arch/powerpc/boot/main.o
```

**Comment by nathanchance** [comment]
For what it's worth, this warning is globally disabled: https://github.com/torvalds/linux/blob/1c7fc5cbc33980acd13d668f1c8f0313d6ae9fd8/Makefile#L704-L705

I've been looking at this for a bit because this is the last Clang warning on a powernv_defconfig build but I am rather confused with what is going on. Where do `_initrd_start` and `_initrd_end` actually get defined? They are declared as `extern` in `arch/powerpc/boot/ops.h` but I don't see where they get defined. It seems obvious that the intention was to check if the value of `_initrd_start` is less than the value of `_initrd_end`, which I assume are addresses but it looks like the actually array addresses are being compared.

**Comment by nickdesaulniers** [comment]
casts to `uintptr_t` from `stdint.h` should help silence the warning.

**Comment by nickdesaulniers** [comment]
@nathanchance looks like you fixed this via series: https://github.com/ClangBuiltLinux/linux/issues?q=label%3A-Wtautological-compare+is%3Aclosed . Last instance?

**Comment by nathanchance** [comment]
Patch sent: https://lore.kernel.org/lkml/20200624035920.835571-1-natechancellor@gmail.com/

I always miss this because `arch/powerpc/boot` does not respect `KCFLAGS=-Werror` which is usually how I catch warnings to fix.

**Comment by nickdesaulniers** [comment]
fixed: df4232d96e724
